src/index.ts

```diff
@@ -47,89 +136,53 @@ export class I18N {
     }
 
     i18n(keysetName: string, key: string, params?: Params): string {
-        const languageData = this.getLanguageData(this.lang);
-
-        if (typeof languageData === 'undefined') {
-            throw new Error(`Language '${this.lang}' is not defined, make sure you call setLang for the same language you called registerKeysets for!`);
+        if (!this.lang && !this.fallbackLang) {
+            throw new Error('Language not specified. You should set at least one of these: "lang", "fallbackLang"');
         }
 
-        if (Object.keys(languageData).length === 0) {
-            this.warn('Language data is empty.');
-
-            return key;
-        }
+        let text: string | undefined;
+        let fallbackText: string | undefined;
+        let details: TranslationData['details'];
 
-        const keyset = languageData[keysetName];
-
-        if (!keyset) {
-            this.warn(
-                'Keyset not found.',
-                keysetName,
-            );
+        ({text, fallbackText, details} = this.getTranslationData({keysetName, key, params, lang: this.lang}));
 
-            return key;
+        if (details && details.code !== ErrorCode.NoLanguageData) {
+            const message = mapErrorCodeToMessage({
+                code: details.code,
+                lang: this.lang,
+                fallbackLang: this.fallbackLang,
+            });
+            this.warn(message, details.keysetName, details.key);
         }
 
-        if (Object.keys(keyset).length === 0) {
-            this.warn(
-                'Keyset is empty.',
-                keysetName,
-            );
-
-            return key;
-        }
-
-        const keyValue = keyset && keyset[key];
-        let result: string;
-
-        if (typeof keyValue === 'undefined') {
-            this.warn(
-                'Missing key.',
+        if (this.fallbackLang) {
+            ({text, fallbackText, details} = this.getTranslationData({
```

Comment:
if `getTranslationData` returns with `details.code === NoLanguageData`, `fallbackText` will be overwritten to `undefined`, right?

Reply:
Yep